File: 001/001.py
import numpy as np

def sum_of_multiples(multiples, max_value):
    """Takes a length 2 list of multiples to check, and an integer maximum
    value to check."""
    # To elegantly solve this problem we are going to find the sum of all the
    # numbers that are divisible by the first number in the list "multiple",
    # and then the second number in the list "multiple". In the process we have
    # over counted by all the numbers that are divisible by the two "multiple"
    # numbers multiplied together. So we subtract out the sum of all the
    # numbers that are divisible by the two "multiple" numbers multiplied
    # together.
    #
    # Explicitly for this problem, we find the sum of numbers divisible by 3,
    # add that to the sum of numbers divisible by 5, and subtract the sum of
    # numbers divisible by 15.
    if len(multiples) != 2:
        print "Expected only two multiples to check."
        print "Out of scope, I'll need additional funds for a scope change."
        return None
    multiples.append(multiples[0]*multiples[1])
    # Make a numpy array for easier math.
    multiples = np.array(multiples)

    # The sum of numbers divisible by a number can be calculated by taking the
    # number itself and adding it to the largest number less than max_value
    # that it is divisible by, and then multiplying it by the half the number
    # of numbers that it is divisible by within the range.
    #
    # Explicitly for this problem for the number 5, there are 999/5 = 199
    # numbers that are divisible by 5 between 1 to 999. The first number is 5
    # and the last number is 999-999%5 = 995. The sum of the first and last
    # number is 1000. The sum of the second number and the second to last
    # number also add to 1000, and so on. Of the 199 numbers that are divisible
    # by 5, there are 199/2.0 of these pairs, and the sum of all these is
    # 1000(199/2.0). NOTE: we divide by a float in case the number of numbers
    # divisible by a number is odd.

    sub_totals = (multiples+max_value-max_value%multiples)*\
            (max_value/multiples)/2.0

    # Sum the first two numbers and subtract the last number to get the total.
    return sum(sub_totals[0:2]) - sub_totals[2]

if __name__ == "__main__":
    print sum_of_multiples(multiples=[3,5], max_value=999)

File: 001/001.c
/***************
Problem 001
Find the sum of all the multiples of 3 or 5 below 1000.
***************/

#include<stdio.h>
#include<math.h>

int solve001( int a, int b, int limit );

void main()
{
    int p1 = 3;
	int p2 = 5;
	int p3 = 1000;
	int ans = solve001(p1, p2, p3-1);
	printf("answer=%i\n",ans);
}

int solve001( int a, int b, int limit )
{
	/**********
	Method:	(3,6,9...999) + (5,10,15...995) - (15,30,45...)
	Separated into three lines for readability:
	(3+6+9+...+999) = 3(1+2+3...333) = 3(1+333)(333/2)
	**********/
	int solve001=0;
	solve001 += a*(1+floor(((float)limit)/a))*floor(((float)limit)/a)/2;
	solve001 += b*(1+floor(((float)limit)/b))*floor(((float)limit)/b)/2;
	solve001 -= (a*b)*(1+floor(((float)limit)/(a*b)))*floor(((float)limit)/(a*b))/2;
	return solve001;
}
